module/zfs/spa.c

```diff
@@ -3327,6 +3339,87 @@ spa_start_livelist_condensing_thread(spa_t *spa)
 	    spa_livelist_condense_cb, spa, minclsyspri);
 }
 
+static int
+spa_rrd_write(spa_t *spa, dmu_tx_t *tx, rrd_t *rrd, const char *what,
+    boolean_t insert)
+{
+
+	if (insert) {
+		return (zap_add(spa_meta_objset(spa), DMU_POOL_DIRECTORY_OBJECT,
+		    what, RRD_ENTRY_SIZE, RRD_STRUCT_ELEM, rrd, tx));
+	}
+
+	return (zap_update(spa_meta_objset(spa),
+	    DMU_POOL_DIRECTORY_OBJECT, what, RRD_ENTRY_SIZE, RRD_STRUCT_ELEM,
+	    rrd, tx));
```

Comment:
IIRC `zap_update()` happily creates new entries if there is no ones already.  So you should not need the explicit `zap_add()` and the `insert` argument.

Reply:
Yeah you are right! Thanks.